fix(x509.c): fixed -checkend return values

snowdroppe opened the pull request:
Fixes #28928 

## Before

```bash
openssl x509 -in /tmp/fullchain.cer -checkend $[3600*24*365*0]; echo $?
Certificate will not expire
0

% openssl x509 -in /tmp/fullchain.cer -checkend $[3600*24*365*10]; echo $?
Certificate will expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*0]; echo $?
Certificate will not expire
Certificate will not expire
Certificate will not expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*1]; echo $?
Certificate will expire
Certificate will not expire
Certificate will not expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*4]; echo $?
Certificate will expire
Certificate will not expire
Certificate will expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*20]; echo $?
Certificate will expire
Certificate will expire
Certificate will expire
0
```

## After

```bash
openssl x509 -in /tmp/fullchain.cer -checkend $[3600*24*365*0]; echo $?
Certificate will not expire
0

% openssl x509 -in /tmp/fullchain.cer -checkend $[3600*24*365*10]; echo $?
Certificate will expire
1

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*0]; echo $?
Certificate will not expire
Certificate will not expire
Certificate will not expire
0

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*1]; echo $?
Certificate will expire
Certificate will not expire
Certificate will not expire
1

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*4]; echo $?
Certificate will expire
Certificate will not expire
Certificate will expire
1

openssl x509 -in /tmp/fullchain.cer -multi -checkend $[3600*24*365*20]; echo $?
Certificate will expire
Certificate will expire
Certificate will expire
1
```

beldmit reviewed (dismissed):
First, I'm not sure it's worth introducing this change.
To distinguish an error because of, say, wrong format, you still need to parse output.

Second, if this change is introduced, we need to document it

snowdroppe commented:
> First, I'm not sure it's worth introducing this change. To distinguish an error because of, say, wrong format, you still need to parse output.

The current documentation reads `Checks if the certificate expires within the next arg seconds and exits non-zero if yes it will expire or zero if not.` which holds for versions prior to the implementation of `-multi` but is uniformly `0` after.

I agree that return codes may mask other errors and am happy to take the organisation's direction on how it should behave.

> 
> Second, if this change is introduced, we need to document it

Valid point as a lot of scripts currently depend on the return code to check for expired / near expired certs. I'll update it according to what is agreed.

Is https://github.com/openssl/openssl/issues/28928 a better thread to discuss the intent?

shahsb commented on apps/x509.c:
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
-        goto end_cert_loop;
+
+        if (multi && k > 0)
+            ret = (ret | expired);
```
Could be done something like following:

```
int expired = X509_cmp_time(X509_get0_notAfter(x), &tcheck) < 0;
ret = expired;  /* For single or first cert */
/* ... print ... */
if (multi && k > 0) {
    ret |= expired;  /* Accumulate */
}
```

snowdroppe replied:
Agree on the shorthand for the bitwise or shorthand assignment.

I disagree on the unconditional assignment of `ret = expired;` as this will execute for all certs in the loop, overwriting previous return codes. This would result in the final return being entirely dependent on the final certificate only.

```c
if (multi && k > 0)
    // 2nd or later cert in chain
    ret |= expired;
else
    // Single cert or 1st in chain
    ret = expired;
```

shahsb replied:
Thanks @snowdroppe!

You're absolutely right -- my earlier unconditional ret = expired; would overwrite on every iteration, leaving only the last cert's status. Excellent catch, and your if/else structure nails it perfectly.

shahsb commented on apps/x509.c:
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
+            ret = expired;
+
+        if (multi && k < sk_X509_num(certs) - 1)
+            goto end_cert_loop;
```
could we please update documentation for this change?

(doc/man1/openssl-x509.pod) claims: "If the certificate will expire within the given number of seconds, then exit code 1 is returned." But it doesn't cover -multi. The PR's change aligns with this but breaks legacy scripts assuming always-0

snowdroppe replied:
Old man page entry:

>Checks if the certificate expires within the next _arg_ seconds and exits nonzero if yes it will expire or zero if not.

Proposed new man page entry:

>Without **-multi** checks if the certificate expires within the next _arg_ seconds and exits non-zero if it will expire or zero if not.
>
>With **-multi** checks if any certificate in the input will expire within the next _arg_ seconds and exits non-zero if any will expire or zero if none will.

shahsb replied:
LGTM. Thanks for making the changes!

shahsb commented on apps/x509.c:
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
+        else
+            ret = expired;
+
+        if (multi && k < sk_X509_num(certs) - 1)
```
Alternatives considered? Could add a new flag (e.g., -strict-checkend), but direct fix is simpler.

t8m replied:
No @shahsb This is a clear regression in 3.6 version. Just needs to be fixed.

shahsb reviewed (commented):
Thanks @snowdroppe for your efforts on this!

It'll LGTM after the suggested fixes!

snowdroppe commented on apps/x509.c:
```diff
@@ -1098,13 +1098,22 @@ int x509_main(int argc, char **argv)
+        if (multi && k < sk_X509_num(certs) - 1)
+            goto end_cert_loop;
+        else
+            goto end;
```
Added

>Check if any certificates in a chain are due to expire within the next 30 days (returns zero if none will expire, non-zero if any will expire):
>
>`openssl x509 -in chain.pem -multi -checkend $[3600*24*30] \`
> `        && echo 'perform renewal' || echo 'renewal unnecessary'`

snowdroppe commented:
Pushed new commit which rebases to master and implements the above suggestions. Pending review, I will rebase again and squash into a single commit.

snowdroppe commented:
I believe the PR is now in a fit state for review.
I have squashed changes into a single commit and rebased to master for easier merging.

Please let me know if there are any additional actions on me. 
Also thanks to everyone for helping me contribute!

snowdroppe commented:
The unit tests trigger an overflow on the checkend `time_t` type i.e. year 2038 bug on some build platforms so I will need to revisit this. I'll update once I have a solution.

openssl-machine commented:
This pull request is ready to merge
